### ISSUE TITLE
[generate] handle attention mask (encoder + cross attention) for encoder decoder models

### ISSUE THREAD

**Issue by eustlb** [huggingface/transformers#36020]
### Feature request

For encoder-decoder autoregressive models attention masks, we follow the convention (cf [here](https://github.com/huggingface/transformers/blob/9d2056f12b66e64978f78a2dcb023f65b2be2108/src/transformers/generation/utils.py#L1958C67-L1959C119)):
`attention_mask` →  applies to the encoder
`decoder_attention_mask` →  applies to the decoder

Moreover, we also go with the following convention concerning modeling code:
1. a `CoolModelNameModel` class that handles the forward pass, inferring the encoder if `encoder_outputs` is not provided and inferring the decoder
2. a usage-specific class that will inherit from the amazing `GenerationMixin` and hold a `CoolModelNameModel` instance. 

For reference, the original code for `SpeechEncoderDecoderModel` forward calls (same goes for other encoder-decoder models like [Bart](https://github.com/huggingface/transformers/blob/9d2056f12b66e64978f78a2dcb023f65b2be2108/src/transformers/models/bart/modeling_bart.py#L1643-L1659) or [Moonshine](https://github.com/huggingface/transformers/blob/9d2056f12b66e64978f78a2dcb023f65b2be2108/src/transformers/models/moonshine/modeling_moonshine.py#L1589-L1603)):
https://github.com/huggingface/transformers/blob/9d2056f12b66e64978f78a2dcb023f65b2be2108/src/transformers/models/speech_encoder_decoder/modeling_speech_encoder_decoder.py#L545-L558
where `encoder_attention_mask` is inferred from the passed `attention_mask`.

🐛 **Such model inputs are prepared in the `prepare_inputs_for_generation`, nevertheless `attention_mask` is lost in:**
https://github.com/huggingface/transformers/blob/9d2056f12b66e64978f78a2dcb023f65b2be2108/src/transformers/generation/utils.py#L409-L411

### Motivation

When doing batch inference and therefore requiring most of the time padding on the inputs values + attention_mask, it is required for the forward of the model to be passed the `attention_mask` in order to compute the correct encoder hidden state and mask the cross attentions.

### Your contribution

I've implemented a naïve fix in #36018 🤗

**Comment by eustlb** [comment]
cc @gante @zucchini-nlp

**Comment by zucchini-nlp** [comment]
Oh right! The fix was needed for Whisper models which needs position ids unlike some other text models. I am wondering, aren't we supposed to prepare encoder-hidden-states before the generation starts, so there should be no need for encoder attention anymore?

**Comment by eustlb** [comment]
Unfortunately Whisper is not a good reference for that! They do not mask the encoder hidden states but rather rely on the fact that the model learned that padding zeros on the input values before passing to the encoder represent silence (so no need to mask). Nevertheless most encoder-decoder models will still need an attention mask for the cross attentions (see Bart or Moonshine for example).
